edgehog-device-runtime-forwarder/src/connection.rs

```diff
@@ -21,37 +21,80 @@
 //! A connection is responsible for sending and receiving data through a WebSocket connection from
 //! and to the [`ConnectionsManager`](crate::connections_manager::ConnectionsManager).
 
-use std::ops::{Deref, DerefMut};
+use std::ops::{ControlFlow, Deref, DerefMut};
 
+use async_trait::async_trait;
 use displaydoc::Display;
+use futures::{SinkExt, StreamExt};
+use http::Request;
 use thiserror::Error as ThisError;
-use tokio::sync::mpsc::Sender;
+use tokio::select;
+use tokio::sync::mpsc::{channel, Receiver, Sender};
 use tokio::task::{JoinError, JoinHandle};
-use tracing::{debug, error, instrument, span, Level};
+use tracing::{debug, error, instrument, span, trace, Level};
+use tungstenite::{Error as TungError, Message as TungMessage};
 
+use crate::connections_manager::WsStream;
 use crate::messages::{
-    Http as ProtoHttp, HttpMessage as ProtoHttpMessage, HttpResponse, Id, ProtoMessage,
-    Protocol as ProtoProtocol, ProtocolError,
+    Http as ProtoHttp, HttpMessage as ProtoHttpMessage, HttpRequest as ProtoHttpRequest,
+    HttpResponse as ProtoHttpResponse, Id, ProtoMessage, ProtocolError,
+    WebSocketMessage as ProtoWebSocketMessage,
 };
 
+pub(crate) const WS_CHANNEL_SIZE: usize = 50;
+
 /// Connection errors.
 #[non_exhaustive]
 #[derive(Display, ThisError, Debug)]
 pub enum ConnectionError {
     /// Channel error.
-    ChannelToWs,
+    Channel(&'static str),
     /// Reqwest error.
-    Reqwest(#[from] reqwest::Error),
+    Http(#[from] reqwest::Error),
     /// Protobuf error.
     Protobuf(#[from] ProtocolError),
     /// Failed to Join a task handle.
     JoinError(#[from] JoinError),
+    /// Message sent to the wrong protocol
+    WrongProtocol,
+    /// Error when receiving message on websocket connection, `{0}`.
+    WebSocket(#[from] TungError),
+    /// Trying to poll while still connecting.
+    Connecting,
+}
+
+#[derive(Debug)]
+pub(crate) enum WriteHandle {
```

Comment:
Because it is used to store the write side of the channel used by the Connections Manager to send WebSocket messages to the respective Connection that will handle it.
Names such as `Connection`, `Protocol` and `Transport` have already been used to define the structs/trait representing a single connection and protobuf internal representation, respectively. `CommunicationProtocol` doesn't reflect the functionality I wanted to expose.